File: app/src/main/java/com/sakuna63/tumbin/application/fragment/PhotoPostFragment.java
package com.sakuna63.tumbin.application.fragment;

import android.graphics.Rect;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.widget.NestedScrollView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.widget.LinearLayout;

import com.hannesdorfmann.fragmentargs.annotation.FragmentWithArgs;
import com.sakuna63.tumbin.application.contract.PostContract;
import com.sakuna63.tumbin.application.widget.GifControlImageView;
import com.sakuna63.tumbin.data.model.Post;
import com.sakuna63.tumbin.databinding.FragmentPhotoPostBinding;

@FragmentWithArgs
public class PhotoPostFragment extends PostFragment
        implements NestedScrollView.OnScrollChangeListener {
    public static final String TAG = PhotoPostFragment.class.getSimpleName();

    PostContract.Presenter presenter;

    private FragmentPhotoPostBinding binding;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        binding = FragmentPhotoPostBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding.scrollView.setOnScrollChangeListener(this);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        presenter.init();
    }

    @Override
    public void showPost(@NonNull Post post) {
        binding.setPost(post);
        ViewTreeObserver viewTreeObserver = binding.containerPhotos.getViewTreeObserver();
        viewTreeObserver.addOnDrawListener(new ViewTreeObserver.OnDrawListener() {
            @Override
            public void onDraw() {
                if (getUserVisibleHint()) {
                    startAnimationGif();
                }
            }
        });
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);

        boolean isViewNotCreated = binding == null;
        if (isViewNotCreated) {
            return;
        }

        if (isVisibleToUser) {
            startAnimationGif();
        } else {
            stopAnimationGif();
        }
    }

    @Override
    public void startAnimationGif() {
        LinearLayout container = binding.containerPhotos;
        Rect screenBounds = new Rect();
        for (int i = 0; i < container.getChildCount(); i++) {
            View child = container.getChildAt(i);
            binding.scrollView.getHitRect(screenBounds);
            if (child instanceof GifControlImageView && child.getLocalVisibleRect(screenBounds)) {
                ((GifControlImageView) child).setRunnable(true);
            }
        }
    }

    @Override
    public void stopAnimationGif() {
        LinearLayout container = binding.containerPhotos;
        for (int i = 0; i < container.getChildCount(); i++) {
            View child = container.getChildAt(i);
            if (child instanceof GifControlImageView) {
                ((GifControlImageView) child).setRunnable(false);
            }
        }
    }

    @Override
    public void setPresenter(PostContract.Presenter presenter) {
        this.presenter = presenter;
    }

    @Override
    public void onScrollChange(NestedScrollView v, int scrollX, int scrollY, int oldScrollX, int oldScrollY) {
        LinearLayout container = binding.containerPhotos;
        Rect screenBounds = new Rect();
        for (int i = 0; i < container.getChildCount(); i++) {
            View child = container.getChildAt(i);
            v.getHitRect(screenBounds);
            if (!(child instanceof GifControlImageView)) {
                continue;
            }

            if (child.getLocalVisibleRect(screenBounds)) {
                ((GifControlImageView) child).setRunnable(true);
            } else {
                ((GifControlImageView) child).setRunnable(false);
            }
        }
    }
}

File: app/src/main/java/com/sakuna63/tumbin/application/fragment/PhotoPostFragment.kt
package com.sakuna63.tumbin.application.fragment

import android.databinding.DataBindingUtil
import android.graphics.Rect
import android.os.Bundle
import android.support.v4.widget.NestedScrollView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.sakuna63.tumbin.application.contract.PhotoPostContract
import com.sakuna63.tumbin.application.contract.presenter.PhotoPostPresenter
import com.sakuna63.tumbin.application.misc.Arg
import com.sakuna63.tumbin.data.dao.DashboardRealmDaoImpl
import com.sakuna63.tumbin.data.model.Post
import com.sakuna63.tumbin.databinding.FragmentPhotoPostBinding
import com.sakuna63.tumbin.databinding.LayoutPhotoBinding
import com.sakuna63.tumbin.extension.applyArguments
import com.sakuna63.tumbin.extension.children
import com.sakuna63.tumbin.extension.put

class PhotoPostFragment : BaseFragment(),
        PhotoPostContract.View, NestedScrollView.OnScrollChangeListener {

    val postId: Long by Arg()

    lateinit private var presenter: PhotoPostContract.Presenter // init on onCreate
    lateinit private var binding: FragmentPhotoPostBinding // init on onCreateView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        presenter = PhotoPostPresenter(postId, this,
                DashboardRealmDaoImpl(activityComponent.realmConfiguration()))
    }

    override fun onCreateView(inflater: LayoutInflater?,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        binding = FragmentPhotoPostBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.scrollView.setOnScrollChangeListener(this)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        presenter.init()
    }

    override fun showPost(post: Post) {
        binding.post = post
        binding.containerPhotos.viewTreeObserver.addOnDrawListener {
            if (userVisibleHint) startVisibleGifAnimation()
        }
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        view ?: return
        presenter.onVisibleToUser(isVisibleToUser)
    }

    override fun startVisibleGifAnimation() {
        val screenBounds = Rect().apply { binding.scrollView.getHitRect(this) }
        binding.containerPhotos.children().forEach {
            val image = DataBindingUtil.getBinding<LayoutPhotoBinding?>(it)?.image
            if (image != null && image.getLocalVisibleRect(screenBounds)) {
                image.isRunnable = true
            }
        }
    }

    override fun stopAllGifAnimation() {
        binding.containerPhotos.children().forEach {
            val image = DataBindingUtil.getBinding<LayoutPhotoBinding?>(it)?.image
            image?.apply { isRunnable = false }
        }
    }

    override fun setPresenter(presenter: PhotoPostContract.Presenter) {
        // no-op
    }

    override fun onScrollChange(v: NestedScrollView, scrollX: Int, scrollY: Int,
                                oldScrollX: Int, oldScrollY: Int) {
        presenter.onScrollChanged()
    }

    companion object {
        val TAG = PhotoPostFragment::class.java.simpleName!!

        fun newInstance(postId: Long) = PhotoPostFragment().applyArguments {
            put(PhotoPostFragment::postId, postId)
        }
    }
}
